### ISSUE TITLE
Menubar, Menu, MenuButton: not accessible

### ISSUE THREAD

**Issue by zhiying99** [primefaces/primefaces#6561]
**Describe the defect**
p:menubar (p:menu, p:meneButton as well) is not accessable. That is, when tabbing through the menu using a screen reader like JAWS, it does not read any of the menu items, even all the generated elements have titles or aria-lables.

**Reproducer**
See here: https://www.primefaces.org/showcase/ui/menu/menubar.xhtml

**To Reproduce**
Open a screen reader (JAWS for example), tab through "File" "Edit" menus and submenus, the screen reader will read nothing.

**Expected behavior**
It would be nice to have an unified focus and screen reader interaction behaviour, like p:tabMenu, which works perfectly.

**Example XHTML**
It seems that the focus is always landed on the surrounding elements like <div ... ></div>, or <li … />, while the accessable values are inside the inner elements. However, simply adding a title or aria-label to <li … /> does not work either.

**Comment by jepsar** [comment]
Relates to #4928

**Comment by melloware** [comment]
@zhiying99 I think this may be a JAWS flaw.  We are following the ARIA guidelines exactly.

For MenuBar see: https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-2/menubar-2.html

```xml
<ul role="menubar" class="ui-menu-list ui-helper-reset">
    <li class="ui-widget ui-menuitem ui-corner-all ui-menu-parent" role="none">
        <a role="menuitem" aria-haspopup="true" href="#" tabindex="-1" class="ui-menuitem-link ui-submenu-link ui-corner-all">
```

You can see the `<ul role="menubar"` and the `<li role="none"` and the item itself has `<a role="menuitem" aria-haspopup="true" `. All of that is according to the ARIA spec.


> tabindex="-1" | a | Makes the menuitem element keyboard focusable but not part of the Tab sequence of the page.
> -- | -- | --

**Comment by melloware** [comment]
@zhiying99 I am going to close this for now unless you can determine exacltly what we are missing in the ARIA spec.  If not we will assume its bug in JAWS interpretation of the HTML.

**Comment by zhiying99** [comment]
I am not sure. the role="menuitem" and aria-haspopup="true" are in li-element instead of in a-element. 
Actually the a-element does not need a tabindex="-1" but in the generated html there it is.  
And the  "div tabindex="0" class="ui-helper-hidden-accessible" makes the screen reader (both JAWS and NVDA, although I am sometimes realy annoyed by JAWS;) silent.

**Comment by melloware** [comment]
Please see the spec. https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-2/menubar-2.html

The role="menuitem" absolutely goes in the `a` element and the tabindex="-1" is also required as part of the spec.


tabindex="-1" | a | Makes the menuitem element keyboard focusable but not part of the Tab sequence of the page.
-- | -- | --
tabindex="0"  | a | Makes the menuitem element keyboard focusable and part of the tab sequence of the page.Only one menuitem in the menubar has tabindex="0".When the page loads, the first item in the menubar has tabindex="0".Focus is managed using roving tabindex.

**Comment by zhiying99** [comment]
Thank for the response. Yes the spec. page https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-2/menubar-2.html works fine (read by Screen Reader). What does not read is here https://www.primefaces.org/showcase/ui/menu/menubar.xhtml.
It's fine for me to close the issue.

**Comment by melloware** [comment]
Hmmm can you check the differences in why it does not read?  I wonder if it tabIndex="0" stuff?

**Comment by zhiying99** [comment]
As descibed in "Example XHTML" part, the tab-focus is landed on the surrounding elements like <div ... >, or <li … />, while the accessable values are inside the inner a-elements. If simply change the tabindex stuff (remove or change), the focus is not visiable any more.  

I am working on the other similar issue, i.e.  tab-focus-contrast on menus. It displayed different background color with hover, even they are defined exactly the same (shown also the same in the element inspector). To be honest, I don't know why.

**Comment by melloware** [comment]
@MarkusChur Can you help us here?  We are trying to figure out why MenuBar for example is not accessible even though it looks like we are following the ARIA spec correctly.

**Comment by melloware** [comment]
will close for now unless someone can figure out what in the spec we are not following but from looking at it we are doing everything correctly.

**Comment by serial389** [comment]
We are facing the same problem. The primefaces menubar is not compatible to JAWS or NVDA.
It seems the primefaces menubar is not handling the menuitem `a` tags for “roving tabindex” the way it is described in the specs:
[https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-2/menubar-2.html](https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-2/menubar-2.html)

For the menuitem `a` tags
tabindex="-1" | a | Makes the menuitem element keyboard focusable but not part of the Tab sequence of the page.
-- | -- | --
tabindex="0" | a | Makes the menuitem element keyboard focusable and part of the tab sequence of the page. Only one menuitem in the menubar has tabindex="0". When the page loads, the first item in the menubar has tabindex="0". `Focus is managed using roving tabindex.`

See the description of “roving tabindex” in the specs:
[https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_roving_tabindex](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_roving_tabindex)

The main difference is, that when navigating to the menubar by pressing the tab key in primefaces, the focus is on a non visible `div` tag, whereas in the menubar from the w3.org website the focus is on the first menuitem `a` tag as described in the specs for “roving tabindex”.

For completeness there are two additional differences between primefaces menubar and the w3.org menubar:
- The aria-expanded attribute is missing on the menuitem `a` tag
- For the submenu the aria-label is missing on the menu `ul` tag

**Comment by melloware** [comment]
@serial389 Thanks for helping.  i think I understand the roving tabindex issue.  However for the `ul` what should the `aria-label` be?

**Comment by serial389** [comment]
I have found this difference by comparing the primefaces menubar with the w3.org menubar tag by tag.
After some tests with JAWS and NVDA it seems that both screen reader are ignoring that `aria-label` attribute for the submenu `ul` tag.
So today for JAWS and NVDA it doesn’t matter if this `aria-label` exist or not.

But this `aria-label` is listed in the spec:

Role | Attribute | Element | Usage
-- | -- | -- | --
menu | aria-label="string" | ul | Defines an accessible name for the menu.

In the w3.org menubar this `aria-label` attribute has the same value as the visible content of the menuitem `a` tag.
For an example.
If the menuitem `a` tag is `<a role="menuitem" aria-haspopup="true" aria-expanded="false" ...>Menu item text</a>`
Then the submenu `ul` tag is: `<ul role="menu" aria-label="Menu item text" ...>`

**Comment by melloware** [comment]
OK submitted a PR. I I think I got all the issues listed!

**Comment by serial389** [comment]
I just gave your pull request a try and I’m sorry to say that it will not fix the problem.
We have tested both JAWS and NVDA and both are not working as expected. 

First of all:
When you are not using a screen reader the primefaces menubar works fine.
But with a running screen reader you cannot navigate within the menubar. The used keys cannot be interpreted correctly, because the system is in an undefined state.

For me the root cause of the problem is the non visible “ui-helper-hidden-accessible” `div` tag that has tabindex=”0” and therefor is in the tab sequence instead of the first menuitem `a` tag.

The screen reader expects the focus to be on the first menuitem.
In this case the screen reader would collect the following information:
- Visible text of the menutiem (for this example “item text”).
- How many sibling menutiem exists (for this example 5).
- What is the number of this current active menutiem (for this example 1).
- Does this menutiem has an attribute aria-haspopup="true" (for this example true).

The screen reader would then say something like: “item text” “menu” “1 out of 5”.

But when the focus is on the non visible “ui-helper-hidden-accessible” `div` tag, the screen reader can’t collect any of the needed information and gets confused.

**Comment by melloware** [comment]
OK can you do me a favor and test regular Menu and MenuButton to see if they are broken the same way or whether the screenreader is working for those?

So it sounds like instead of giving focus to “ui-helper-hidden-accessible” we have to find the very first menu item  and make that tabindex=0 instead of that hidden accessible.

**Comment by melloware** [comment]
Also can you tell me that if once the MenuBar does get focus and we start setting tabindex=0 to the highlighted menu item does the screenreader start working then?  It looks like a log of this code for MenuBar and MegaMenu is assuming it has this "ui-helper-hidden-accessible" keyboard trap.

**Comment by melloware** [comment]
@serial389 Can you pull my PR branch again I added a fix to remove tabindex off the ui-helper-hidden-accessible once the menu has focus and returns it to it after you leave the menu.

**Comment by melloware** [comment]
Two things 1. Don't forget to CTRL+F5 to reload JavaScript and 2. I wonder if the keyboard trap DIV should have aria-hidden=true?

**Comment by melloware** [comment]
OK I added aria-role="none" on the keyboard target div which should not read it out loud for assitive technolgoies but should allow you to TAB to it and once the Menu has focus we assign the first menu option tabindex=0 so hopefully the screenreader picks that up?

**Comment by serial389** [comment]
I just tested your last pull request and the keyboard target `div` tag is still in the tab sequence.
When you tab into the menubar the focus is still first on the keyboard target `div` tag.
The job of a screen reader is to tell the user what he can do on this web element when it has focus.
But since in this case it is a `div` tag the screen reader will only say something like “tab able web element”
You can imagine that this will not help the blind user much.

But when you press the tab key again the focus will now be on the first menuitem.
Both screen readers will detect this fine and say what is expected for this first menuitem.

To your request to test the both Menu and MenuButton on primefaces showcase.
They do not work as expected because the navigation within does not work (pressing the up and down arrow keys).
Both screen reader show a different behavior.

For your last test we run the pull request code and have edited the html code in the browser to add the `aria-hidden=”true”` attribute to the keyboard target `div` tag.
Now the two screen reader behaves different but not as expected.
NVDA: Reads the whole menubar content for all existing menuitems.
JAWS: Reads only the first menuitem but in a wrong way (The submenu status is wrong)

**Comment by melloware** [comment]
@serial389 Can you pull my latest PR I added role"none" to the keyboard trap which is supposed to tell all screenreaders to ignore it.

**Comment by melloware** [comment]
`role="none"` is preferred over `aria-hidden="true"`

**Comment by melloware** [comment]
I just tested myself.  Seems like no matter what this keyboard trap is a problem.  May have to rethink the way these menus are done.

**Comment by Groell-Philipp** [comment]
The problem here, and in other widgets as well, is that for keyboard navigation inside a widget focus management must happen programatically. A mix of tabindex and ARIA role/property attributes alone won't do the trick. See section 6 of the WAI-ARIA authoring practices.

You can find working examples of both vertical and horizontal menus at

* https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-1/menubar-1.html
* https://www.w3.org/TR/wai-aria-practices-1.1/examples/menu-button/menu-button-actions.html.

For the vertical menus (Menu, MenuButton, TieredMenu etc.) I've got a working solution as this is also part of my thesis. So if you'd like me to, I can share that once my thesis is submited and graded.

**Comment by melloware** [comment]
Yes please share or submit a PR for review when you are ready!

**Comment by ZeroBrushV2** [comment]
Hello, any update about this fix?

**Comment by sojoio** [comment]
Also interested to know if this ever got fixes as I am having the same problem

**Comment by melloware** [comment]
Nope the ticket is still open.  @Groell-Philipp was going to submit a PR but never did.

**Comment by melloware** [comment]
Just for everyone monitoring this thread thanks to @Rene-Stoeckel I am making great progress on fixing this forever.    Its a lot of changes but they are all for making this more accessible, screen reader friendly, and keyboard friendly!

**Comment by melloware** [comment]
OK this is a massive change but anyone out there with screenreaders that would like to test and give feedback i updated Menu, MenuButton, MenuBar, and TieredMenu.

See: http://20.161.11.140/showcase/ui/menu/menubar.xhtml

Poke around that BETA showcase and let me know if the screenreader is working. so far internal feedback has been positive but I know there are many different screenreaders out there.

**Comment by melloware** [comment]
I have received feedback that ORCA and NVDA are reading it correctly. Just waiting on feedback about JAWS or any others.

**Comment by Rene-Stoeckel** [comment]
We have checked the Beta showcase mentioned above with JAWS 2024. The menu navigation with the screenreader works using cursor keys and it is reading all menu items correctly. Looks very promising!

**Comment by NarutoTI** [comment]
Maybe, if there was an attribut like preserveState="true/false" it would help.
The important thing is the screen readers we tested are reading really nicely in the PF menu. So I don't think it is a big issue for now.

**Comment by melloware** [comment]
OK I deployed a new version that remembers the last focused root item: http://20.161.11.140/showcase/ui/menu/menubar.xhtml

I did this because that is what the ARIA example does.

**Comment by Rene-Stoeckel** [comment]
Our team took a look again on the new showcase and confirms good interoperatibility with NVDA and JAWS. 
These 3 minor findings were yet observed:
•	When the fokus is on the most left menu item and the cursor left key is pressed the fokus indicator disappears, but it should stay visible.
•	When a sub-submenu is opened and it is then closed with the left cursor key, the whole menu is closed after 1 or 2 seconds.
•	The arrow-icons in the menu entries are reported as part of the menu entry label by the screenreader (as an unnown Character). They should be hidden with a aria-hidden=true attribute to avoid it.

**Comment by melloware** [comment]
@Rene-Stoeckel i have issue 1 and 3 fixed but as for the sub-submenu closing the whole menu after 1 or 2 seconds i am having trouble reproducing?  This is on the MenuBar example right?

**Comment by melloware** [comment]
Makes sense I am using Firefox.  Let me try Chrome.

**Comment by melloware** [comment]
OK all 3 items addressed: http://20.161.11.140/showcase/ui/menu/menubar.xhtml

Let me know how it goes

**Comment by Rene-Stoeckel** [comment]
Yes, looks good, the 3 issues are fixed.
